### PR TITLE
code-server: init code-server-module

### PR DESCRIPTION
**Pull request by stackshadow**
###### Motivation for this change

- an nixos-module was missing for code-server to easy config code-server

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/#sec-conf-file))
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [21.11 Release Notes (or backporting 21.05 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2111-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [x] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/doc/manual/release-notes/rl-2111.section.md [hunk 1]
```diff
@@ -60,6 +60,10 @@ subsonic-compatible api. Available as [navidrome](#opt-services.navidrome.enable
 
 - [sx](https://github.com/earnestly/sx), a simple alternative to both xinit and startx for starting a Xorg server. Available as [services.xserver.displayManager.sx](#opt-services.xserver.displayManager.sx.enable)
 
+- `code-server`-module now available
+
+- `code-server-bin`-package is now available with code-server version 3.11.1
+
```
Comment (SuperSandro2000) [inline nixos/doc/manual/release-notes/rl-2111.section.md:66]:
```suggestion
- `code-server`-module now available

```

File: pkgs/servers/code-server-bin/default.nix [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+    zip
+    makeWrapper
+    git
+    rsync
+    jq
+    moreutils
+  ];
+
+  configurePhase = "true";
+  buildPhase = "true";
```
Comment (SuperSandro2000) [inline pkgs/servers/code-server-bin/default.nix:35]:
```suggestion
  dontConfigure = true;
  dontBuild = true;
```

File: pkgs/servers/code-server-bin/default.nix [hunk 3]
```diff
@@ -0,0 +1,54 @@
...
+  installPhase = ''
+    ${rsync}/bin/rsync -av ./ $out
```
Comment (SuperSandro2000) [inline pkgs/servers/code-server-bin/default.nix:37]:
```suggestion
    rsync -av ./ $out
```

File: pkgs/servers/code-server-bin/default.nix [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+    rm $out/lib/node
+    ln -fs ${pkgs.nodejs}/bin/node $out/lib/node
+  '';
+
+  meta = with lib; {
+    description =
+      "Run VS Code on any machine anywhere and access it in the browser.";
```
Comment (SuperSandro2000) [inline pkgs/servers/code-server-bin/default.nix:44]:
```suggestion
    description = "Run VS Code on any machine anywhere and access it in the browser.";
```

**Review by SuperSandro2000 — CHANGES_REQUESTED** [review]

File: pkgs/servers/code-server-bin/default.nix (hunk 4 above)
Comment (stackshadow) [reply pkgs/servers/code-server-bin/default.nix:44]: Which formatter do you use ? I use nixfmt to format my *.nix files, i just wondering why nixfmt do an newline here :thinking:

File: nixos/modules/services/web-apps/code-server.nix [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+      package = mkOption {
+        default = pkgs.code-server;
+        defaultText = "pkgs.code-server";
+        description = "Which code-server derivation to use";
+        type = types.package;
+      };
+
+      packages = mkOption {
+        default = [ ];
+        defaultText = "[]";
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:21]: I don't think you need that here... 🤔

File: nixos/modules/services/web-apps/code-server.nix [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        description = "Packages that are available in the PATH of code-server";
+        example = "[ pkgs.go ]";
+        type = types.listOf types.package;
+      };
+
+      host = mkOption {
+        default = "127.0.0.1";
+        defaultText = "Locahost-IP";
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:29]: Drop this line please.

File: nixos/modules/services/web-apps/code-server.nix [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+        description = "The host-ip to bind to";
+        type = types.str;
+      };
+
+      port = mkOption {
+        default = "4444";
+        defaultText = "4444";
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:36]: Redundant.

File: nixos/modules/services/web-apps/code-server.nix [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+        description = "The port where code-server runs";
+        type = types.str;
+      };
+
+      auth = mkOption {
+        default = "password";
+        defaultText = "Use password";
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:43]: Drop this line.

File: nixos/modules/services/web-apps/code-server.nix [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+        description = "The type of authentication to use. [password, none]";
+        type = types.str;
+      };
+
+      hashedPassword = mkOption {
+        default = "";
+        defaultText = "The hashed password, need auth set to 'password'";
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:50]: Drop this line.

File: nixos/modules/services/web-apps/code-server.nix [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+        description = "Create the password with: 'echo -n 'thisismypassword' | npx argon2-cli -e'";
+        type = types.str;
+      };
+
+      user = mkOption {
+        default = "code-server";
+        defaultText = "The internal code-server user";
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:57]: Drop this line.

File: nixos/modules/services/web-apps/code-server.nix [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+        description = "Set the username for code-server. You can set this to your username to access your own files";
+        type = types.str;
+      };
+
+      groups = mkOption {
+        default = "[]";
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:63]: You don't want to quote this.

File: nixos/modules/services/web-apps/code-server.nix [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+        defaultText = "[]";
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:64]: Drop this line.

File: nixos/modules/services/web-apps/code-server.nix [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+        description = "An array of additional groups for the code-server user";
+        example = ''[ "docker" ]'';
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:66]: Maybe you want to use `literalExample` here... 🤔

File: nixos/modules/services/web-apps/code-server.nix [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+        type = types.listOf types.str;
+      };
+
+    };
+  };
+
+  ###### implementation
+  config = mkIf config.services.code-server.enable {
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:74]:
```suggestion
  config = mkIf cfg.enable {
```

File: nixos/modules/services/web-apps/code-server.nix [hunk 15]
```diff
@@ -0,0 +1,109 @@
...
+    systemd.services.code-server = let
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:75]: Please drop the unused `let`...`in`.

File: nixos/modules/services/web-apps/code-server.nix [hunk 16]
```diff
@@ -0,0 +1,109 @@
...
+
+    in {
+      description = "VSCode server";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network-online.target" ];
+      path = cfg.packages;
+      environment = {
+        HASHED_PASSWORD = "${cfg.hashedPassword}";
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:83]: Quotes not needed.

File: nixos/modules/services/web-apps/code-server.nix [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+        PKG_CONFIG_PATH = "/run/current-system/sw/lib/pkgconfig";
+      };
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/code-server --disable-telemetry --bind-addr ${cfg.host}:${cfg.port} --auth ${cfg.auth}";
+        ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+        RuntimeDirectory = "${cfg.package}";
+        User = "${cfg.user}";
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:90]: Quotes not required.

File: nixos/modules/services/web-apps/code-server.nix [hunk 18]
```diff
@@ -0,0 +1,109 @@
...
+    };
+
+    users.users.code-server =
+      mkIf (config.services.code-server.user == "code-server") {
+        isNormalUser = true;
+        createHome = true;
+        description = "code-server user";
+        home = "/home/code-server";
+        extraGroups = cfg.groups;
+        shell = pkgs.zsh;
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:106]: My guess is that `bash` is probably more compatible.

File: nixos/modules/services/web-apps/code-server.nix [hunk 19]
```diff
@@ -0,0 +1,109 @@
...
+        Restart = "on-failure";
+        # for ping probes
+        AmbientCapabilities = [ "CAP_NET_RAW" ];
+
+      };
+
+    };
+
+    users.users.code-server =
+      mkIf (config.services.code-server.user == "code-server") {
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:100]:
```suggestion
      mkIf (cfg.user == "code-server") {
```

File: nixos/modules/services/web-apps/code-server.nix [hunk 20]
```diff
@@ -0,0 +1,109 @@
...
+        User = "${cfg.user}";
+        Restart = "on-failure";
+        # for ping probes
+        AmbientCapabilities = [ "CAP_NET_RAW" ];
+
+      };
+
+    };
+
+    users.users.code-server =
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:99]: Please ensure a group exists as well.

File: nixos/modules/services/web-apps/code-server.nix [hunk 21]
```diff
@@ -0,0 +1,109 @@
...
+        HASHED_PASSWORD = "${cfg.hashedPassword}";
+        PKG_CONFIG_PATH = "/run/current-system/sw/lib/pkgconfig";
+      };
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/code-server --disable-telemetry --bind-addr ${cfg.host}:${cfg.port} --auth ${cfg.auth}";
+        ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+        RuntimeDirectory = "${cfg.package}";
+        User = "${cfg.user}";
+        Restart = "on-failure";
+        # for ping probes
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:92]: Can you please elaborate?

File: nixos/modules/services/web-apps/code-server.nix [hunk 22]
```diff
@@ -0,0 +1,109 @@
...
+    systemd.services.code-server = let
+
+    in {
+      description = "VSCode server";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network-online.target" ];
+      path = cfg.packages;
+      environment = {
+        HASHED_PASSWORD = "${cfg.hashedPassword}";
+        PKG_CONFIG_PATH = "/run/current-system/sw/lib/pkgconfig";
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:84]: Why?

File: nixos/modules/services/web-apps/code-server.nix [hunk 23]
```diff
@@ -0,0 +1,109 @@
...
+      };
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/code-server --disable-telemetry --bind-addr ${cfg.host}:${cfg.port} --auth ${cfg.auth}";
+        ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+        RuntimeDirectory = "${cfg.package}";
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:89]: I don't think this is correct.

File: nixos/modules/services/web-apps/code-server.nix [hunk 24]
```diff
@@ -0,0 +1,109 @@
...
+        # for ping probes
+        AmbientCapabilities = [ "CAP_NET_RAW" ];
+
+      };
+
+    };
+
+    users.users.code-server =
+      mkIf (config.services.code-server.user == "code-server") {
+        isNormalUser = true;
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:101]: This option implies `createHome` and `home` so you can remove them.

**Review by aanderse — CHANGES_REQUESTED** [review]
Hello. I left a review on your module. If you need any help or have any questions please don't hesitate to ask. I'm always happy to lend a helping hand.

File: nixos/modules/services/web-apps/code-server.nix [hunk 25]
```diff
@@ -0,0 +1,109 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+let cfg = config.services.code-server;
+
+in {
+  ###### interface
+  options = {
+    services.code-server = {
+      enable = mkEnableOption "enable code-server";
```
[the remark this answers is not in the document]
Comment (stackshadow) [reply nixos/modules/services/web-apps/code-server.nix:10]: Ah, okay, right, i read it in the docs :) and change this

File: nixos/modules/services/web-apps/code-server.nix [hunk 26]
```diff
@@ -0,0 +1,109 @@
...
+
+      package = mkOption {
+        default = pkgs.code-server;
+        defaultText = "pkgs.code-server";
+        description = "Which code-server derivation to use";
```
[the remark this answers is not in the document]
Comment (stackshadow) [reply nixos/modules/services/web-apps/code-server.nix:15]: Got it :+1:

File: nixos/modules/services/web-apps/code-server.nix (hunk 5 above)
Comment (stackshadow) [reply nixos/modules/services/web-apps/code-server.nix:21]: Well, its really neede for usage of code-server.
Problem is, that code-server is started by systemd. Lets say we would like to develop in go ( i do this in my day to day job ) problem is, that code-server is started by systemd. In the env of systemd there is no PATH to the `go` binary, even if its installed as system-package in nixos. Thats why i use the `path` option in the systemd-service. 
I was thinking if i can use `environment.systemPackages` for that :thinking:

File: nixos/modules/services/web-apps/code-server.nix [hunk 27]
```diff
@@ -0,0 +1,109 @@
...
+        defaultText = "4444";
+        description = "The port where code-server runs";
+        type = types.str;
+      };
+
+      auth = mkOption {
+        default = "password";
+        defaultText = "Use password";
+        description = "The type of authentication to use. [password, none]";
+        type = types.str;
```
[the remark this answers is not in the document]
Comment (stackshadow) [reply nixos/modules/services/web-apps/code-server.nix:45]: Oh cool, i was not aware of the enum-type, thanks :+1:

File: nixos/modules/services/web-apps/code-server.nix (hunk 15 above)
Comment (stackshadow) [reply nixos/modules/services/web-apps/code-server.nix:75]: Ah damn, it was from previous tests

File: nixos/modules/services/web-apps/code-server.nix (hunk 22 above)
Comment (stackshadow) [reply nixos/modules/services/web-apps/code-server.nix:84]: Ah, history, i was forced to use imagemagick with golang and pkg-config is not enough to install, but i will convert this to an option :)

File: nixos/modules/services/web-apps/code-server.nix (hunk 23 above)
Comment (stackshadow) [reply nixos/modules/services/web-apps/code-server.nix:89]: Jep, better to run in the users-home

File: nixos/modules/services/web-apps/code-server.nix (hunk 18 above)
Comment (stackshadow) [reply nixos/modules/services/web-apps/code-server.nix:106]: Right, better not to use thinks that i personally prefer :)

File: nixos/modules/services/web-apps/code-server.nix (hunk 21 above)
Comment (stackshadow) [reply nixos/modules/services/web-apps/code-server.nix:92]: This is a web service and i run it currently on a server. If it crash, it should auto-restart the service on failure.

Comment (aanderse) [reply nixos/modules/services/web-apps/code-server.nix:92]: `Restart` makes perfect sense. I meant `AmbientCapabilities`.

Comment (stackshadow) [reply nixos/modules/services/web-apps/code-server.nix:115]: I check if code-server work without these settings, i think its an copy-paste-misuse :laughing:

Comment (stackshadow) [reply nixos/modules/services/web-apps/code-server.nix:115]: Okay, ping works without :+1:  i remove this setting

File: nixos/modules/services/web-apps/code-server.nix (hunk 20 above)
Comment (stackshadow) [reply nixos/modules/services/web-apps/code-server.nix]: Jep, right, default group will be created if needed :+1:

File: nixos/modules/services/web-apps/code-server.nix (hunk 22 above)
Comment (stackshadow) [reply nixos/modules/services/web-apps/code-server.nix]: Done, env-vars for code-server can now be added by `extra-environment`-option

File: nixos/modules/services/web-apps/code-server.nix (hunk 23 above)
Comment (stackshadow) [reply nixos/modules/services/web-apps/code-server.nix]: Done, i set it to the username like in similar services

**Comment by aanderse** [comment]
@stackshadow I have pushed a few commits to a branch which you can feel free to include in your PR here: https://github.com/stackshadow/nixpkgs/compare/code-server-bin...aanderse:nixos/code-server?expand=1

Let me know if you have any questions.

**Comment by stackshadow** [comment]
> @stackshadow I have pushed a few commits to a branch which you can feel free to include in your PR here: https://github.com/stackshadow/nixpkgs/compare/code-server-bin...aanderse:nixos/code-server?expand=1
> 
> Let me know if you have any questions.

Cool thanks, its now more readable :)

File: nixos/modules/services/web-apps/code-server.nix [hunk 28]
```diff
@@ -0,0 +1,130 @@
...
+        description =
+          "Create the password with: 'echo -n 'thisismypassword' | npx argon2-cli -e'.";
+        type = types.str;
+      };
+
+      user = mkOption {
+        default = defaultUser;
+        example = "yourUser";
+        description = ''
+          The user to run Syncthing as.
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:67]: Oops! Copy+paste gone wrong with "Syncthing".

File: nixos/modules/services/web-apps/code-server.nix [hunk 29]
```diff
@@ -0,0 +1,130 @@
...
+      path = cfg.extraPackages;
+      environment = {
+        HASHED_PASSWORD = cfg.hashedPassword;
+      } // cfg.extraEnvironment;
+      serviceConfig = {
+        ExecStart =
+          "${cfg.package}/bin/code-server --disable-telemetry --bind-addr ${cfg.host}:${toString cfg.port} --auth ${cfg.auth}";
+        ExecReload = "${pkgs.coreutils}/bin/kill -HUP $MAINPID";
+        RuntimeDirectory = cfg.user;
+        User = cfg.user;
```
Comment (aanderse) [inline nixos/modules/services/web-apps/code-server.nix:109]:
```suggestion
        User = cfg.user;
        Group = cfg.group;
```

**Review by aanderse — CHANGES_REQUESTED** [review]
Just a few more small things. Also, please squash this down to 2 commits:
- adding yourself to the maintainers list
- adding the module

Thanks!

File: nixos/modules/services/web-apps/code-server.nix (hunk 28 above)
Comment (stackshadow) [reply nixos/modules/services/web-apps/code-server.nix]: Done :+1:

File: nixos/modules/services/web-apps/code-server.nix (hunk 29 above)
Comment (stackshadow) [reply nixos/modules/services/web-apps/code-server.nix:108]: Added, thanks for mentioned it :)

**Comment by stackshadow** [comment]
> Just a few more small things. Also, please squash this down to 2 commits:
> 
>     * adding yourself to the maintainers list
> 
>     * adding the module
> 
> 
> Thanks!

Was a little bit tricky to split the commit into two, but this train my git muscle :D

**Comment by aanderse** [comment]
Looking good. You forgot to actually set yourself as the maintainer, though.

File: nixos/modules/services/web-apps/code-server.nix [hunk 30]
```diff
@@ -0,0 +1,130 @@
...
+    systemd.services.code-server = {
+      description = "VSCode server";
+      wantedBy = [ "multi-user.target" ];
+      after = [ "network-online.target" ];
+      path = cfg.extraPackages;
+      environment = {
+        HASHED_PASSWORD = cfg.hashedPassword;
+      } // cfg.extraEnvironment;
+      serviceConfig = {
+        ExecStart = "${cfg.package}/bin/code-server --disable-telemetry --bind-addr ${cfg.host}:${toString cfg.port} --auth ${cfg.auth}";
```
Comment (contrun) [inline nixos/modules/services/web-apps/code-server.nix:105]: Please add an option to let users fill in extra arguments.

**Comment by stackshadow** [comment]
> Looking good. You forgot to actually set yourself as the maintainer, though.

Ahh right, done :D

File: nixos/modules/services/web-apps/code-server.nix (hunk 30 above)
Comment (stackshadow) [reply nixos/modules/services/web-apps/code-server.nix]: Done, thanks for the idea :+1:

**Review by aanderse — APPROVED** [review]
LGTM :+1: Any reason not to merge? Pull the trigger?

**Comment by GoogleBot42** [comment]
Is this ready to merge?

**Comment by stackshadow** [comment]
I've rebased the branch to fix the conflicts :)

**Comment by stackshadow** [comment]
So for me its ready to merge :)

**Comment by aanderse** [comment]
Thanks everyone!

**Pull request merged**